pdns/xsk.cc

```diff
@@ -82,20 +82,20 @@ struct UmemEntryStatus
   Status status{Status::Free};
 };
 
-LockGuarded<std::unordered_map<uint64_t, UmemEntryStatus>> s_umems;
+LockGuarded<std::map<std::pair<void*, uint64_t>, UmemEntryStatus>> s_umems;
 
-void checkUmemIntegrity(const char* function, int line, uint64_t offset, const std::set<UmemEntryStatus::Status>& validStatuses, UmemEntryStatus::Status newStatus)
+void checkUmemIntegrity(const char* function, int line, std::shared_ptr<LockGuarded<vector<uint64_t>>> vect, uint64_t offset, const std::set<UmemEntryStatus::Status>& validStatuses, UmemEntryStatus::Status newStatus)
 {
   auto umems = s_umems.lock();
-  if (validStatuses.count(umems->at(offset).status) == 0) {
-    std::cerr << "UMEM integrity check failed at " << function << ": " << line << ": status is " << static_cast<int>(umems->at(offset).status) << ", expected: ";
+  if (validStatuses.count(umems->at({vect.get(), offset}).status) == 0) {
+    std::cerr << "UMEM integrity check failed at " << function << ": " << line << ": status of " << (void*)vect.get() << ", " << offset << " is " << static_cast<int>(umems->at({vect.get(), offset}).status) << ", expected: ";
     for (const auto status : validStatuses) {
       std::cerr << static_cast<int>(status) << " ";
     }
     std::cerr << std::endl;
     abort();
   }
-  (*umems)[offset].status = newStatus;
+  (*umems)[{vect.get(), offset}].status = newStatus;
```

Comment:
I suggest using `insert` or `emplace` and check return value as it would avoid the the `count` call.

Reply:
I pushed a new commit avoiding the double-lookup when UMEM state debugging is enabled. Note that we are not inserting there, the entry is always present, the `count()` checks if the status is valid.

Reply:
> I pushed a new commit avoiding the double-lookup when UMEM state debugging is enabled. Note that we are not inserting there, the entry is always present, the `count()` checks if the status is valid.

Right, I had the wrong type in mind.